tests/test_utils_cog.py

```diff
@@ -0,0 +1,155 @@
+import pytest
+from pathlib import Path
+import numpy as np
+from types import SimpleNamespace
+from dask.delayed import Delayed
+import dask
+
+from datacube.testutils import (
+    mk_test_image,
+    gen_tiff_dataset,
+)
+from datacube.testutils.io import native_load, rio_slurp_xarray, rio_slurp
+from datacube.utils.cog import write_cog, to_cog, _write_cog
+
+
+def gen_test_data(prefix, dask=False):
+    w, h, dtype, nodata, ndw = 96, 64, 'int16', -999, 7
+
+    aa = mk_test_image(w, h, dtype, nodata, nodata_width=ndw)
+
+    ds, gbox = gen_tiff_dataset(
+        SimpleNamespace(name='aa', values=aa, nodata=nodata), prefix)
+    extras = {}
+
+    if dask:
+        extras.update(dask_chunks={'time': 1})
+
+    xx = native_load(ds, **extras)
+
+    return xx.aa.isel(time=0), ds
+
+
+def test_cog_file(tmpdir):
+    pp = Path(str(tmpdir))
+    xx, ds = gen_test_data(pp)
+
+    # write to file
+    ff = write_cog(xx, pp / "cog.tif")
+    assert isinstance(ff, Path)
+    assert ff == pp / "cog.tif"
+    assert ff.exists()
+
+    yy = rio_slurp_xarray(pp / "cog.tif")
+    np.testing.assert_array_equal(yy.values, xx.values)
+    assert yy.geobox == xx.geobox
+    assert yy.nodata == xx.nodata
+
+    _write_cog(np.stack([xx.values, xx.values]),
+               xx.geobox,
+               pp / "cog-2-bands.tif",
+               overview_levels=[])
+
+    yy, mm = rio_slurp(pp / "cog-2-bands.tif")
+    assert mm.gbox == xx.geobox
+    assert yy.shape == (2, *xx.shape)
+    np.testing.assert_array_equal(yy[0], xx.values)
+    np.testing.assert_array_equal(yy[1], xx.values)
+
+    with pytest.raises(ValueError, match="Need 2d or 3d ndarray on input"):
+        _write_cog(xx.values.ravel(), xx.geobox, pp / "wontwrite.tif")
+
+
+def test_cog_file_dask(tmpdir):
```

Comment:
[tmp_path](https://docs.pytest.org/en/latest/tmpdir.html#the-tmp-path-fixture) is the replacement for `tmpdir`.

Reply:
I think we tried that before, didn't work on 3.5